rdata/conversion/to_r.py

```diff
@@ -75,269 +141,453 @@ def build_r_object(
-            (for example, convert_to_r_object).
+        data: Non-empty list of values or (key, value) pairs.
 
     Returns:
         R object.
     """
-    if convert_value is None:
-        convert_value = convert_to_r_object
-
     if len(data) == 0:
         msg = "data must not be empty"
         raise ValueError(msg)
 
-    if isinstance(data, dict):
-        data = data.copy()
-        key = next(iter(data))
-        value1 = convert_value(data.pop(key), encoding=encoding)
-        tag = build_r_sym(key, encoding=encoding)
-    elif isinstance(data, list):
-        value1 = convert_value(data[0], encoding=encoding)
-        data = data[1:]
+    head = data[0]
+    tail = data[1:]
+    if isinstance(head, tuple):
+        tag, car = head
+    else:
         tag = None
+        car = head
 
-    if len(data) == 0:
-        value2 = build_r_object(RObjectType.NILVALUE)
-    else:
-        value2 = build_r_list(data, encoding=encoding, convert_value=convert_value)
+    cdr = build_r_object(RObjectType.NILVALUE) if len(tail) == 0 else build_r_list(tail)
 
-    return build_r_object(
-        RObjectType.LIST,
-        value=(value1, value2),
-        tag=tag,
-        )
+    return build_r_object(RObjectType.LIST, value=(car, cdr), tag=tag)
 
 
-def build_r_sym(
-        data: str,
-        *,
-        encoding: Encoding,
-) -> RObject:
+class ConverterFromPythonToR:
     """
-    Build R object representing symbol.
+    Class converting Python objects to R objects.
 
-    Args:
-        data: String.
+    Attributes:
         encoding: Encoding to be used for strings within data.
-
-    Returns:
-        R object.
+        format_version: File format version.
+        r_version_serialized: R version written as the creator of the object.
     """
-    r_type = RObjectType.SYM
-    r_value = convert_to_r_object(data.encode(encoding), encoding=encoding)
-    return build_r_object(r_type, value=r_value)
+    def __init__(self, *,
+            encoding: Encoding = "utf-8",
+            format_version: int = DEFAULT_FORMAT_VERSION,
+            r_version_serialized: int = DEFAULT_R_VERSION_SERIALIZED,
+    ) -> None:
+        """
+        Init class.
+
+        Args:
+            encoding: Encoding to be used for strings within data.
+            format_version: File format version.
+            r_version_serialized: R version written as the creator of the object.
+        """
+        self.encoding = encoding
+        self.format_version = format_version
+        self.r_version_serialized = r_version_serialized
+        self._references: dict[str | None, tuple[int, RObject | None]] \
+            = {None: (0, None)}
+
+        # In test files the order in which dataframe attributes are written varies.
+        # R can read files with attributes in any order, but this variable
+        # is used in tests to change the attribute order to match with the test file.
+        self.df_attr_order: list[str] | None = None
```

Comment:
It would be good to know what affects the order.

In absence of that information I would propose to always write them in the most common order, and try to ignore the order of the attributes when comparing the original and created file. If that is not possible, regenerating the old test files is also doable.